dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/EventRequestToSearchParamsMapper.java

```diff
@@ -127,66 +134,35 @@ void setSchema()
-        }
-
-        String orgUnit = eventCriteria.getOrgUnit();
-        OrganisationUnit ou = organisationUnitService.getOrganisationUnit( orgUnit );
-        if ( !StringUtils.isEmpty( orgUnit ) && ou == null )
-        {
-            throw new IllegalQueryException( "Org unit is specified but does not exist: " + orgUnit );
-        }
+        OrganisationUnit orgUnit = organisationUnitService.getOrganisationUnit( eventCriteria.getOrgUnit() );
+        validateOrgUnit( eventCriteria.getOrgUnit(), orgUnit );
 
         User user = currentUserService.getCurrentUser();
-        if ( pr != null && !user.isSuper() && !aclService.canDataRead( user, pr ) )
-        {
-            throw new IllegalQueryException( "User has no access to program: " + pr.getUid() );
-        }
+        validateUser( user, program, programStage );
 
-        if ( ps != null && !user.isSuper() && !aclService.canDataRead( user, ps ) )
-        {
-            throw new IllegalQueryException( "User has no access to program stage: " + ps.getUid() );
-        }
-
-        String trackedEntityInstance = eventCriteria.getTrackedEntity();
-        TrackedEntityInstance tei = entityInstanceService.getTrackedEntityInstance( trackedEntityInstance );
-        if ( !StringUtils.isEmpty( trackedEntityInstance ) && tei == null )
-        {
-            throw new IllegalQueryException(
-                "Tracked entity instance is specified but does not exist: " + trackedEntityInstance );
-        }
+        TrackedEntityInstance trackedEntityInstance = entityInstanceService
+            .getTrackedEntityInstance( eventCriteria.getTrackedEntity() );
+        validateTrackedEntity( eventCriteria.getTrackedEntity(), trackedEntityInstance );
 
         CategoryOptionCombo attributeOptionCombo = inputUtils.getAttributeOptionCombo(
             eventCriteria.getAttributeCc(),
             eventCriteria.getAttributeCos(),
             true );
-        if ( attributeOptionCombo != null && !user.isSuper()
-            && !aclService.canDataRead( user, attributeOptionCombo ) )
-        {
-            throw new IllegalQueryException(
-                "User has no access to attribute category option combo: " + attributeOptionCombo.getUid() );
-        }
+        validateAttributeOptionCombo( attributeOptionCombo, user );
 
-        Set<String> eventIds = eventCriteria.getEvents();
-        validateFilter( eventIds, eventCriteria.getFilter(), programStage, ps );
+        Set<String> eventIds = parseUids( eventCriteria.getEvent() );
+        validateFilter( eventCriteria.getFilter(), eventIds, eventCriteria.getProgramStage(), programStage );
 
-        if ( eventIds == null )
-        {
-            eventIds = new HashSet<>();
-        }
+        Set<String> assignedUserIds = parseUids( eventCriteria.getAssignedUser() );
+        validateAssignedUsers( eventCriteria.getAssignedUserMode(), assignedUserIds );
+
+        EventSearchParams params = new EventSearchParams();
```

Comment:
I wonder if we can manage to not instantiate EventSearchParams and  maybe use a builder later in the code.
To do so we will need to change how we add filters, filterAttributes and dataElements

Reply:
Agree 👍 We'll explore it tomorrow 😊